src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,419 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.DiaSymReader.Tools;
+using Microsoft.Tools.Common;
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Binding;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Xml;
+
+namespace Microsoft.Diagnostics.Tools.Stack
+{
+    internal static class SymbolicateHandler
+    {
+        // Temporary folder to store the files converted from pdb to xml
+        private static readonly string tempDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+
+        delegate void SymbolicateDelegate(FileInfo[] inputPath, string searchDir, string output);
+
+        /// <summary>
+        /// Get the line number from the Method Token and IL Offset at the stacktrace
+        /// </summary>
+        /// <param name="inputPath">The input path for file with stacktrace text</param>
+        /// <param name="searchDir">All paths in the directory to the assembly and pdb where the exception occurred</param>
+        /// <param name="output">The output path for the extracted line number data</param>
+        /// <returns></returns>
+        private static void Symbolicate(FileInfo[] inputPath, string searchDir, string output)
+        {
+            try
+            {
+                if (inputPath == null)
+                {
+                    throw new InvalidDataException("Required argument missing for command\n");
+                }
+
+                string input = string.Empty;
+                if (inputPath.Length > 1)
+                {
+                    for (int i = 0; i < inputPath.Length; i++)
+                    {
+                        Console.WriteLine($" {i + 1}. {inputPath[i]}");
+                    }
+                    Console.Write($"Select one of several stacktrace files [1-{inputPath.Length}]: ");
+                    input = inputPath[Int32.Parse(Console.ReadLine()) - 1].FullName;
+                }
```

Comment:
If a user specifies multiple input files, should the tool symbolicate all of them rather than forcing user interaction to pick one?

It occurred to me that we'd need to do an in-place update of the existing file or drop and updated version of each file next to it. It _also_ occurred to me that we don't have a specific file extension we can look for. Specifying multiple files with a wildcard would be problematic due to that.

@noahfalk / @shirhatti what are your thoughts on handling multiple files? Should we have users script around the tool to hit multiple files?

Reply:
I think it would be better to receive only one specified extension(```xxxxx.stack``` or ```xxxxx.stacktrace```).
What do you think?

Reply:
The issue with specifying an extension is that there isn't an accepted, standard extension for this type of output. This means we either need to have our users manually change the name of each file they want touched or specifically list all the files they want touched. In either case we are requiring the user to manually list out their files.

I'm inclined to not allow wildcards and only accept a list of specific files since it is simpler for our tool to handle. If a user wants to wildcard, they can rely on the shell's ability to do so, e.g., `$ find ./*_exception_log | xargs dotnet-stack symbolicate`.

We will have different default behaviors depending on the number of input files as a result:

* 1 input file --> print modified file to stdout
* \>1 input file --> print modified file to new file next to old one (e.g., input.txt -> intput.sym.txt)

Reply:
> If a user specifies multiple input files, should the tool symbolicate all of them rather than forcing user interaction to pick one?

Should we allow multiple stacktrace files to be input? Isn't it better to input only one file at a time?
Or how about guiding the format of the input file contents?
```
  ------- Input file format -------
   at {typeName}.{methodName}({parameters}) in {moduleName}: token {methodToken}+{ilOffset}
```

Reply:
> @noahfalk / @shirhatti what are your thoughts on handling multiple files? Should we have users script around the tool to hit multiple files?

I would keep it simple (at least initially) and support a single file input and write the output to some path adjacent to the output. Then as  optional improvements we could add arguments that change the output path or write the output to stdout. File output as the default appealed to me because:
a) It requires less shell scripting knowledge
b) Probably if the input started in a file the user wants the output in a file
c) If we later add support for multiple files the experience doesn't need to abruptly change behavior

> Or how about guiding the format of the input file contents?

I assume we'll have to have some definition of the format the tool can parse. I'm not sure if I understood the question though.

Reply:
I concur with @noahfalk. The tool should take one file in, and write to a new file next to the old one, e.g.,

```sh
$ ls
input.txt
$ dotnet stack symbolicate input.txt
$ ls
input.txt	input.symbolicated.txt
```

or whatever addition to the file extension we want to use to differentiate.

If you're asking whether we should require the input file to have a specific format, I'm not sure that's something we can require of users. There isn't a standardized file format that we can expect. I think it is sufficient to look for the `at ... in ... token ...` pattern and fixup when we find it.

Reply:
> There isn't a standardized file format that we can expect.

I think the format of the input file is already set in the function:
https://github.com/dotnet/runtime/blob/c93bb62e33934c3b8b6b1d293612d44360483bd8/src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs#L206
So, the user using this tool thought that it was easy to input without changing the output stack.
Because it is already output as below.
```"   at {typeName}.{methodName}({parameters}) in {moduleName}: token {methodToken}+{ilOffset}"```

> I think it is sufficient to look for the `at ... in ... token ...` pattern and fixup when we find it.

Also, I agree with you. But wouldn't most users use it without change? :)

Reply:
I think we might be saying the same thing.

I think we should take in _any_ file and treat it as a text file because there is no agreed upon file extension for stack traces like this. We should just scan the file for the pattern and attempt to fix it up in place, then save the fixed-up file next to the original.